### PR TITLE
Fix CenterLoss and add metric learning code.

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
1. Fix `CenterLoss` bug, retrain on Mnist dataset
image

3. adapt engine code for metric learning.
4. add visualization for metric learning when execute evaluation.

### CODE REVIEW

File: ppcls/arch/__init__.py [hunk 1]
```diff
@@ -78,15 +78,35 @@ def __init__(self, **config):
...
             y = self.head(x, label)
-            out["logits"] = y
+            if isinstance(y, Tensor):
+                out["logits"] = y
+            elif isinstance(y, dict):
+                out.update(y)
+            else:
+                raise NotImplementedError(
+                    f"head must return dict or Tensor, but got ({type(y)})")
         return out
```
Comment (RainFrost1) [inline ppcls/arch/__init__.py:110]: 尽量使用Dataset、Dataloader的输出控制数据类型，保证各模块的输入统一。保证主模块逻辑简单、清晰

File: ppcls/arch/backbone/model_zoo/lenet.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                    bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
+                    uniform_ = Uniform(-bound, bound)
+                    uniform_(m.bias)
+
+        self.apply(_init_func)
+
+    def forward(self, x):
+        x = self.prelu1_1(self.conv1_1(x))
+        x = self.prelu1_2(self.conv1_2(x))
+        x = F.max_pool2d(x, 2)
```
Comment (RainFrost1) [inline ppcls/arch/backbone/model_zoo/lenet.py:126]: 在模型的结构定义中，尽量使用`nn`模块下的函数及方法，保证repo格式统一吧

File: ppcls/arch/backbone/model_zoo/lenet.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        x = self.prelu3_1(self.conv3_1(x))
+        x = self.prelu3_2(self.conv3_2(x))
+        x = F.max_pool2d(x, 2)
+
+        x = x.reshape([-1, 128 * 3 * 3])
+        x = self.prelu_fc1(self.fc1(x))  # neck, backbone->features
+        y = self.fc2(x)  # head features->logits
+
+        return {'features': x, 'logits': y}
```
Comment (RainFrost1) [inline ppcls/arch/backbone/model_zoo/lenet.py:140]: model应该只有一个返回值，两个返回值的情况，可以参考`Neck、Head`的实现情况

File: ppcls/engine/engine.py [hunk 4]
```diff
@@ -221,10 +220,28 @@ def __init__(self, config, mode="train"):
                     self.model, self.config["Global"]["pretrained_model"])
 
         # build optimizer
+        use_metric_optimizer = False
```
Comment (RainFrost1) [inline ppcls/engine/engine.py:223]: 多个 optimizer的情况，建议更通用的写法，需要讨论一下

File: ppcls/engine/evaluation/classification.py [hunk 5]
```diff
@@ -40,6 +40,12 @@ def classification_eval(engine, epoch_id=0):
         dataset) if not engine.use_dali else engine.eval_dataloader.size
     max_iter = len(engine.eval_dataloader) - 1 if platform.system(
     ) == "Windows" else len(engine.eval_dataloader)
+
+    visualize = engine.config["Global"].get("visualize", False)
+    if visualize:
+        all_features = []
+        all_labels = []
+
```
Comment (RainFrost1) [inline ppcls/engine/evaluation/classification.py:48]: 这块代码其他的实验应该用不到，建议放到复现实验特定的代码中，下同

File: ppcls/loss/centerloss.py [hunk 6]
```diff
@@ -1,54 +1,75 @@
...
-from __future__ import division
-from __future__ import print_function
+from __future__ import absolute_import, division, print_function
+
+from typing import Dict
+
 import paddle
 import paddle.nn as nn
-import paddle.nn.functional as F
+from paddle import Tensor
```
Comment (RainFrost1) [inline ppcls/loss/centerloss.py:7]: loss的输入也保证统一吧
